Nil pointer reference on azd up (Buildpacks)

I was recently working on https://github.com/savannahostrowski/savannah.dev/tree/main and accidentally deleted my Dockerfile. When I ran `azd up`, I got this error. I'm assuming it's because Buildpacks don't have a Go target so it just sort of freaked out? Once I added the Dockerfile back, things were all good.

That said, we should probably catch this and throw something more human-readable.

```
azd deploy --debug
2023/11/12 03:42:58 main.go:54: azd version: 1.4.4 (commit 45a48a32c9f3caad481343869ade05221df26ec7)
2023/11/12 03:42:58 project.go:113: Reading project from file '/workspaces/terminal-personal-site/azure.yaml'
2023/11/12 03:42:58 main.go:208: using cached latest version: 1.4.3 (expires on: 2023-11-12T22:13:12Z)
2023/11/12 03:42:58 cobra_builder.go:141: Resolved action 'azd-deploy-action'
2023/11/12 03:42:58 middleware.go:124: running middleware 'debug'
2023/11/12 03:42:58 middleware.go:124: running middleware 'experimentation'
2023/11/12 03:42:58 experimentation.go:42: assignment context: initflow_t:78271;0g5ad841:76970;
2023/11/12 03:42:58 middleware.go:124: running middleware 'telemetry'
2023/11/12 03:42:58 telemetry.go:50: TraceID: c2d91e82452420b886ebbb4af86ba635
2023/11/12 03:42:58 middleware.go:124: running middleware 'hooks'
2023/11/12 03:42:58 hooks.go:156: service 'web' does not require any command hooks.
2023/11/12 03:42:58 hooks.go:85: azd project is not available or does not contain any command hooks, skipping command hook registrations.
2023/11/12 03:42:58 command_runner.go:307: Run exec: 'docker --version' , exit code: 0
-------------------------------------stdout-------------------------------------------
Docker version 24.0.7-1, build afdd53b4e341be38d2056a42113b938559bb1d94
2023/11/12 03:42:58 docker.go:254: docker version: Docker version 24.0.7-1, build afdd53b4e341be38d2056a42113b938559bb1d94
2023/11/12 03:42:58 docker.go:185: determining version from docker --version string: Docker version 24.0.7-1, build afdd53b4e341be38d2056a42113b938559bb1d94
2023/11/12 03:42:58 docker.go:197: extracted docker version: 24.0.7-1, build: afdd53b4e341be38d2056a42113b938559bb1d94 from version string
2023/11/12 03:42:58 ensure.go:47: Skipping install check for 'Docker'. It was previously confirmed.

Deploying services (azd deploy)

  |       | Deploying service web2023/11/12 03:42:58 framework_service_docker.go:157: building image for service web, cwd: /workspaces/terminal-personal-site, path: ./Dockerfile, context: ., buildArgs: [])
  |      =| Deploying service web (Building Docker image from source)2023/11/12 03:42:58 command_runner.go:307: Run exec: '/home/vscode/.azd/bin/pack --version' , exit code: 0
-------------------------------------stdout-------------------------------------------
0.30.0+git-c38f7da.build-4952
2023/11/12 03:42:58 pack.go:130: pack version: 0.30.0
2023/11/12 03:42:58 pack.go:144: using local pack: /home/vscode/.azd/bin/pack
2023/11/12 03:42:58 appdetect.go:246: Detecting projects in directory: /workspaces/terminal-personal-site
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x5608439de60d]

goroutine 32 [running]:
[github.com/azure/azure-dev/cli/azd/pkg/project.(*dockerProject).packBuild(0xc00047ff80,](http://github.com/azure/azure-dev/cli/azd/pkg/project.(*dockerProject).packBuild(0xc00047ff80,) {0x56084443eb48, 0xc000332e70}, 0xc000368ea0, {{0x560843cc080c, 0xc}, {0x560843e55538, 0x1}, {0x560843cbf56b, 0xb}, ...}, ...)
        /mnt/vss/_work/1/s/cli/azd/pkg/project/framework_service_docker.go:323 +0x46d
[github.com/azure/azure-dev/cli/azd/pkg/project.(*dockerProject).Build.func1(0xc0000bb920)](http://github.com/azure/azure-dev/cli/azd/pkg/project.(*dockerProject).Build.func1(0xc0000bb920))
        /mnt/vss/_work/1/s/cli/azd/pkg/project/framework_service_docker.go:182 +0x748
[github.com/azure/azure-dev/cli/azd/pkg/async.(*TaskWithProgress[...]).Run.func1()](http://github.com/azure/azure-dev/cli/azd/pkg/async.(*TaskWithProgress[...]).Run.func1())
        /mnt/vss/_work/1/s/cli/azd/pkg/async/task.go:144 +0xfd
created by [github.com/azure/azure-dev/cli/azd/pkg/async.(*TaskWithProgress[...]).Run](http://github.com/azure/azure-dev/cli/azd/pkg/async.(*TaskWithProgress[...]).Run) in goroutine 31
        /mnt/vss/_work/1/s/cli/azd/pkg/async/task.go:140 +0x1c5
```

--- a/cli/azd/pkg/project/framework_service_docker.go
+++ b/cli/azd/pkg/project/framework_service_docker.go
@@ -25,6 +25,7 @@ import (
 	"github.com/azure/azure-dev/cli/azd/pkg/input"
 	"github.com/azure/azure-dev/cli/azd/pkg/output"
 	"github.com/azure/azure-dev/cli/azd/pkg/tools"
+	"github.com/azure/azure-dev/cli/azd/pkg/tools/azcli"
 	"github.com/azure/azure-dev/cli/azd/pkg/tools/docker"
 	"github.com/azure/azure-dev/cli/azd/pkg/tools/pack"
 	"go.opentelemetry.io/otel/trace"
@@ -270,12 +271,16 @@ func (p *dockerProject) Package(
 const DefaultBuilderImage = "mcr.microsoft.com/oryx/builder:debian-bullseye-20231004.1"
 const DefaultDotNetBuilderImage = "mcr.microsoft.com/oryx/builder:debian-buster-20231004.1"
 
+// All buildpacks groups have failed to detect w/o error.
+// See https://buildpacks.io/docs/concepts/components/lifecycle/detect/#exit-codes
+const PackExitCodeUndetected = 20
+
 func (p *dockerProject) packBuild(
 	ctx context.Context,
 	svc *ServiceConfig,
 	dockerOptions DockerProjectOptions,
 	imageName string) (*ServiceBuildResult, error) {
-	pack, err := pack.NewPackCli(ctx, p.console, p.commandRunner)
+	packCli, err := pack.NewPackCli(ctx, p.console, p.commandRunner)
 	if err != nil {
 		return nil, err
 	}
@@ -369,7 +374,7 @@ func (p *dockerProject) packBuild(
 		)
 	}
 
-	err = pack.Build(
+	err = packCli.Build(
 		ctx,
 		svc.Path(),
 		builder,
@@ -379,6 +384,18 @@ func (p *dockerProject) packBuild(
 	p.console.StopPreviewer(ctx, false)
 	if err != nil {
 		span.EndWithStatus(err)
+
+		var statusCodeErr *pack.StatusCodeError
+		if errors.As(err, &statusCodeErr) && statusCodeErr.Code == pack.StatusCodeUndetectedNoError {
+			return nil, &azcli.ErrorWithSuggestion{
+				Err: err,
+				Suggestion: "Failed to build image from source without a Dockerfile. " +
+					fmt.Sprintf(
+						"To provide explicit instructions for building the image, author a Dockerfile and save it as %s",
+						filepath.Join(svc.Path(), dockerOptions.Path)),
+			}
+		}
+
 		return nil, err
 	}
 

--- a/cli/azd/pkg/project/framework_service_docker.go
+++ b/cli/azd/pkg/project/framework_service_docker.go
@@ -271,10 +271,6 @@ func (p *dockerProject) Package(
 const DefaultBuilderImage = "mcr.microsoft.com/oryx/builder:debian-bullseye-20231004.1"
 const DefaultDotNetBuilderImage = "mcr.microsoft.com/oryx/builder:debian-buster-20231004.1"
 
-// All buildpacks groups have failed to detect w/o error.
-// See https://buildpacks.io/docs/concepts/components/lifecycle/detect/#exit-codes
-const PackExitCodeUndetected = 20
-
 func (p *dockerProject) packBuild(
 	ctx context.Context,
 	svc *ServiceConfig,

--- a/cli/azd/pkg/project/framework_service_docker.go
+++ b/cli/azd/pkg/project/framework_service_docker.go
@@ -320,23 +320,25 @@ func (p *dockerProject) packBuild(
 				return nil, err
 			}
 
-			for _, dep := range prj.Dependencies {
-				if dep == appdetect.PyFastApi {
-					launch, err := appdetect.PyFastApiLaunch(prj.Path)
-					if err != nil {
-						return nil, err
-					}
+			if prj != nil {
+				for _, dep := range prj.Dependencies {
+					if dep == appdetect.PyFastApi {
+						launch, err := appdetect.PyFastApiLaunch(prj.Path)
+						if err != nil {
+							return nil, err
+						}
 
-					// If launch isn't detected, fallback to default Oryx runtime logic, which may recover for scenarios
-					// such as a simple main entrypoint launch.
-					if launch != "" {
-						environ = append(environ,
-							"POST_BUILD_COMMAND=pip install uvicorn",
-							//nolint:lll
-							"ORYX_RUNTIME_SCRIPT=oryx create-script -appPath ./oryx-output -bindPort 80 -userStartupCommand "+
-								"'uvicorn "+launch+" --port $PORT --host $HOST' && ./run.sh")
+						// If launch isn't detected, fallback to default Oryx runtime logic, which may recover for scenarios
+						// such as a simple main entrypoint launch.
+						if launch != "" {
+							environ = append(environ,
+								"POST_BUILD_COMMAND=pip install uvicorn",
+								//nolint:lll
+								"ORYX_RUNTIME_SCRIPT=oryx create-script -appPath ./oryx-output -bindPort 80 -userStartupCommand "+
+									"'uvicorn "+launch+" --port $PORT --host $HOST' && ./run.sh")
+						}
+						break
 					}
-					break
 				}
 			}
 		}

--- a/cli/azd/pkg/project/framework_service_docker.go
+++ b/cli/azd/pkg/project/framework_service_docker.go
@@ -318,33 +318,13 @@ func (p *dockerProject) packBuild(
 				))
 		}
 
-		// Support for FastAPI apps since the Oryx builder does not support it yet
 		if svc.Language == ServiceLanguagePython {
-			prj, err := appdetect.DetectDirectory(ctx, svc.Path())
+			pyEnviron, err := getEnvironForPython(ctx, svc)
 			if err != nil {
 				return nil, err
 			}
-
-			if prj != nil {
-				for _, dep := range prj.Dependencies {
-					if dep == appdetect.PyFastApi {
-						launch, err := appdetect.PyFastApiLaunch(prj.Path)
-						if err != nil {
-							return nil, err
-						}
-
-						// If launch isn't detected, fallback to default Oryx runtime logic, which may recover for scenarios
-						// such as a simple main entrypoint launch.
-						if launch != "" {
-							environ = append(environ,
-								"POST_BUILD_COMMAND=pip install uvicorn",
-								//nolint:lll
-								"ORYX_RUNTIME_SCRIPT=oryx create-script -appPath ./oryx-output -bindPort 80 -userStartupCommand "+
-									"'uvicorn "+launch+" --port $PORT --host $HOST' && ./run.sh")
-						}
-						break
-					}
-				}
+			if len(pyEnviron) > 0 {
+				environ = append(environ, pyEnviron...)
 			}
 		}
 	}
@@ -389,9 +369,9 @@ func (p *dockerProject) packBuild(
 		if errors.As(err, &statusCodeErr) && statusCodeErr.Code == pack.StatusCodeUndetectedNoError {
 			return nil, &azcli.ErrorWithSuggestion{
 				Err: err,
-				Suggestion: "Failed to build image from source without a Dockerfile. " +
+				Suggestion: "No Dockerfile was found, and image could not be automatically built from source. " +
 					fmt.Sprintf(
-						"To provide explicit instructions for building the image, author a Dockerfile and save it as %s",
+						"\nSuggested action: Author a Dockerfile and save it as %s",
 						filepath.Join(svc.Path(), dockerOptions.Path)),
 			}
 		}
@@ -414,6 +394,41 @@ func (p *dockerProject) packBuild(
 			ImageName: imageName,
 		},
 	}, nil
+}
+
+func getEnvironForPython(ctx context.Context, svc *ServiceConfig) ([]string, error) {
+	prj, err := appdetect.DetectDirectory(ctx, svc.Path())
+	if err != nil {
+		return nil, err
+	}
+
+	if prj == nil { // Undetected project, resume build from the Oryx builder
+		return nil, nil
+	}
+
+	// Support for FastAPI apps since the Oryx builder does not support it yet
+	for _, dep := range prj.Dependencies {
+		if dep == appdetect.PyFastApi {
+			launch, err := appdetect.PyFastApiLaunch(prj.Path)
+			if err != nil {
+				return nil, err
+			}
+
+			// If launch isn't detected, fallback to default Oryx runtime logic, which may recover for scenarios
+			// such as a simple main entrypoint launch.
+			if launch == "" {
+				return nil, nil
+			}
+
+			return []string{
+				"POST_BUILD_COMMAND=pip install uvicorn",
+				//nolint:lll
+				"ORYX_RUNTIME_SCRIPT=oryx create-script -appPath ./oryx-output -bindPort 80 -userStartupCommand " +
+					"'uvicorn " + launch + " --port $PORT --host $HOST' && ./run.sh"}, nil
+		}
+	}
+
+	return nil, nil
 }
 
 func getDockerOptionsWithDefaults(options DockerProjectOptions) DockerProjectOptions {

--- a/cli/azd/pkg/tools/pack/pack.go
+++ b/cli/azd/pkg/tools/pack/pack.go
@@ -12,7 +12,9 @@ import (
 	"net/http"
 	"os"
 	"path/filepath"
+	"regexp"
 	"runtime"
+	"strconv"
 	"strings"
 
 	"github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"
@@ -29,6 +31,28 @@ import (
 // PackVersion is the minimum version of pack that we require (and the one we fetch when we fetch pack on behalf of a
 // user).
 var PackVersion semver.Version = semver.MustParse("0.30.0")
+
+var statusCodeFailureRegexp = regexp.MustCompile(`failed with status code: (\d+)`)
+
+// All buildpacks groups have failed to detect w/o error.
+// See https://buildpacks.io/docs/concepts/components/lifecycle/detect/#exit-codes
+const StatusCodeUndetectedNoError = 20
+
+// StatusCodeError is a status code error provided by pack CLI.
+type StatusCodeError struct {
+	Err error
+
+	// See all available status codes https://buildpacks.io/docs/concepts/components/lifecycle/create/
+	Code int
+}
+
+func (s *StatusCodeError) Error() string {
+	return s.Err.Error()
+}
+
+func (s *StatusCodeError) Unwrap() error {
+	return s.Err
+}
 
 type PackCli interface {
 	Build(
@@ -200,12 +224,31 @@ func (cli *packCli) Build(
 		runArgs = runArgs.WithStdOut(progressWriter).WithStdErr(progressWriter)
 	}
 
-	_, err = cli.runner.Run(ctx, runArgs)
+	res, err := cli.runner.Run(ctx, runArgs)
 	if err != nil {
-		return err
+		return wrapStatusCodeErr(err, res)
 	}
 
 	return nil
+}
+
+func wrapStatusCodeErr(err error, res exec.RunResult) error {
+	if err == nil {
+		return err
+	}
+
+	matches := statusCodeFailureRegexp.FindStringSubmatch(res.Stderr)
+	if len(matches) == 2 {
+		code, parseErr := strconv.Atoi(matches[1])
+		if parseErr == nil {
+			return &StatusCodeError{
+				Err:  err,
+				Code: code,
+			}
+		}
+	}
+
+	return err
 }
 
 func packName() string {

